File: app/src/main/java/com/wonium/cicada/ui/activity/MainActivity.java
/*
 * Copyright (C) 2019 Ethan.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.wonium.cicada.ui.activity;


import com.alibaba.android.arouter.facade.annotation.Route;
import com.wonium.cicada.R;
import com.wonium.cicada.databinding.ActivityMainBinding;
import com.wonium.cicada.router.PageRouter;
import com.wonium.cicada.ui.fragment.AccountFragment;
import com.wonium.cicada.ui.fragment.FindFragment;
import com.wonium.cicada.ui.fragment.FriendFragment;
import com.wonium.cicada.ui.fragment.MyFragment;
import com.wonium.cicada.ui.fragment.VideoFragment;
import com.wonium.cicada.base.BaseActivity;

import java.util.ArrayList;
import java.util.List;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;

/**
 * @ClassName: MainActivity
 * @Description: 添加描述
 * @Author: Wonium
 * @E-mail: wonium@qq.com
 * @Blog: https://blog.wonium.com
 * @CreateDate: 2018/11/28 14:14
 * @UpdateUser: 添加更新者
 * @UpdateDate: 2018/11/28 14:14
 * @UpdateDescription: 更新描述
 * @Version:
 */
@Route(path = PageRouter.ACTIVITY_MAIN)
public class MainActivity extends BaseActivity {
    private ActivityMainBinding mBinding;
    private List<Fragment> fragments;

    @Override
    public void initWindowAttributes() {
        setAllowFullScreen(false);
    }

    @Override
    public int getLayoutResId() {
        return R.layout.activity_main;
    }

    @Override
    public void bindLayout(int layoutResId) {
        mBinding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(mBinding.getRoot());
    }

    @Override
    public void initView() {
        // 开启状态栏沉浸
        setStatusBar(true);
        mBinding.includeToolbar.tvToolbarTitle.setText(getString(R.string.app_name));
        initFragment();
    }

    /**
     * 设置状态栏颜色
     * @return
     */
    @Override
    protected int getStatusColor () {
        return getResources().getColor(R.color.black);
    }

    @Override
    public void initListener() {
        mBinding.includeMainContent.bottomNavigationView.setOnNavigationItemSelectedListener(menuItem -> {
            int itemId;
            switch (menuItem.getItemId()) {
                case R.id.nav_find:
                    itemId =0;
                    break;
                case R.id.nav_video:
                    itemId=1;
                    break;
                case R.id.nav_my:
                    itemId =2;
                    break;
                case R.id.nav_friend:
                    itemId=3;
                    break;
                case R.id.nav_account:
                    itemId=4;
                    break;
                default:
                    itemId=0;
                    break;
            }
            replaceFragment(lastShowFragment,itemId);
            return true;
        });
    }
    private int lastShowFragment=0;

    private void initFragment() {

        fragments = new ArrayList<>();
        fragments.add(FindFragment.newInstance("FindFragment", "0"));
        fragments.add(VideoFragment.newInstance("VideoFragment", "1"));
        fragments.add(MyFragment.newInstance("MyFragment","2"));
        fragments.add(FriendFragment.newInstance("FriendFragment","3"));
        fragments.add(AccountFragment.newInstance("AccountFragment","4"));
        getSupportFragmentManager()
                .beginTransaction()
                .add(mBinding.includeMainContent.container.getId(),fragments.get(0))
                .show(fragments.get(0))
                .commitAllowingStateLoss();
    }

    private  void replaceFragment(int lastShowFragment,int index){
        FragmentTransaction transaction =getSupportFragmentManager().beginTransaction();
        transaction.hide(fragments.get(lastShowFragment));
        this.lastShowFragment =index;
        if (!fragments.get(index).isAdded()){
            transaction.add(mBinding.includeMainContent.container.getId(),fragments.get(index));
        }
        transaction.show(fragments.get(index)).commitAllowingStateLoss();
    }



}

File: app/src/main/java/com/ethan/hydrogen/demo/ui/activity/MainActivity.kt
/*
 * Copyright (C) 2019 Ethan.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.ethan.hydrogen.demo.ui.activity

import android.content.Context
import android.content.Intent
import android.view.MenuItem
import com.ethan.hydrogen.demo.R
import com.ethan.hydrogen.demo.base.BaseActivity
import com.ethan.hydrogen.demo.databinding.ActivityMainBinding
import com.ethan.hydrogen.demo.ui.fragment.AccountFragment
import com.ethan.hydrogen.demo.ui.fragment.FragmentTools
import com.ethan.hydrogen.demo.ui.fragment.FriendFragment
import com.ethan.hydrogen.demo.ui.fragment.MyFragment
import com.ethan.hydrogen.demo.ui.fragment.VideoFragment
import com.ethan.hydrogen.ui.BaseFragment
import com.ethan.hydrogen.utils.ResourceUtil

/**
 * @author ethan
 */
class MainActivity : BaseActivity() {

    private val mBinding: ActivityMainBinding by lazy { ActivityMainBinding.inflate(layoutInflater) }
    private val fragments: MutableList<BaseFragment> by lazy {
        mutableListOf<BaseFragment>().apply {
            add(FragmentTools.newInstance(args1 = "FragmentTools", args2 = "0"))
            add(VideoFragment.newInstance("VideoFragment", "1"))
            add(MyFragment.newInstance("MyFragment", "2"))
            add(FriendFragment.newInstance("FriendFragment", "3"))
            add(AccountFragment.newInstance("AccountFragment", "4"))
        }

    }

    override fun initWindowAttributes() {
        setAllowFullScreen(false)
    }

    override fun getLayoutResId(): Int {
        return R.layout.activity_main
    }

    override fun bindLayout(layoutResId: Int) {
        setContentView(mBinding.root)
    }

    override fun initView() { // 开启状态栏沉浸
        setStatusBar(true)
        mBinding.includeToolbar.tvToolbarTitle.text = getString(R.string.app_name)
        initFragment()
    }

    /**
     * 设置状态栏颜色
     *
     * @return
     */
    override fun getStatusColor(): Int {
        return ResourceUtil.getInstance()
            .getColor(context, com.ethan.hydrogen.R.color.black)
    }

    override fun initListener() {
        mBinding.includeMainContent.bottomNavigationView.setOnNavigationItemSelectedListener { menuItem: MenuItem ->
            replaceFragment(
                lastShowFragment, when (menuItem.itemId) {
                    R.id.nav_tools -> 0
                    R.id.nav_view  -> 1
                    R.id.nav_libs  -> 2
                    R.id.nav_mine  -> 3
                    else           -> 0
                }
            )
            true
        }
    }

    private var lastShowFragment = 0
    private fun initFragment() {
        supportFragmentManager.beginTransaction()
            .add(mBinding.includeMainContent.container.id, fragments[0])
            .show(fragments[0])
            .commitAllowingStateLoss()
    }

    private fun replaceFragment(lastShowFragment: Int, index: Int) {
        val transaction = supportFragmentManager.beginTransaction()
        transaction.hide(fragments[lastShowFragment])
        this.lastShowFragment = index
        if (fragments[index].isAdded.not()) {
            transaction.add(mBinding.includeMainContent.container.id, fragments[index])
        }
        transaction.show(fragments[index])
            .commitAllowingStateLoss()
    }

    companion object {
        @JvmStatic
        fun start(context: Context) {
            context.startActivity(Intent(context, MainActivity::class.java))
        }
    }
}